_posts/2021-04-05-reactive-mode.md

```diff
@@ -0,0 +1,150 @@
+</div>
+
+Let's discuss the components:
+
+**Flink**
+
+- The **JobManager** is deployed as a [Kubernetes job](https://kubernetes.io/docs/concepts/workloads/controllers/job/). We are submitting a container that is based on the official Flink Docker image, but has the jar file of our job added to it. The Flink job simply reads data from a Kafka topic and does some expensive math operations per event received. We use these math operations to generate high CPU loads, without requiring a large Kafka deployment.
+- The **TaskManager(s)** are deployed as a Kubernetes deployment, which is scaled through a [Horizontal Pod Autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/). In this experiment, the autoscaler is monitoring the CPU load of the pods in the deployment. The number of pods is adjusted between 1 and 15 pods by the autoscaler.
+
+**Additional Components**:
+
+- We have a **Zookeeper** and **Kafka** deployment (each with one pod) to provide a Kafka topic that serves as the input for the Flink job.
+- **Data Generator** pod which produces messages to the Kafka topic. The data generator is generating simple string messages at a adjustable rate. In this experiment, the rate is following a sine wave.
+- For monitoring, we are deploying **Prometheus** and **Grafana**.
+
+The entire setup is [available on GitHub](https://github.com/rmetzger/flink-reactive-mode-k8s-demo) if you want to try this out yourself.
+
+### Results
+
+We've deployed above components on a hosted Kubernetes cluster, running it for several days. The results are best examined based on a simple Grafana dashboard:
+
+<div class="row front-graphic">
+  <img src="{{ site.baseurl }}/img/blog/2021-04-reactive-mode/result.png" alt="Reactive Mode Demo Result"/>
+  <p class="align-center">Reactive Mode Experiment Results</p>
+</div>
+
+Let's take a look at the dashboard:
+
+- On the top left, we see the **Kafka consumer lag**, reported by Flink's Kafka consumer (source). This metric reports basically the queue size of unprocessed messages. A high lag means that Flink is not processing messages as fast as they are produced: we need to scale up. 
+
+  In the chart you see that the lag is usually following the throughput of data coming from Kafka. When the throughput is the highest, the reported lag is at \~75k messages. In low throughput times, it is basically at zero.
+
+- On the top right, you see the **throughput**, measured in Records per second, as reported by Flink. You see that the throughput is roughly following a sine wave, peaking at 6k messages per second, and going down to almost zero.
+
+- The bottom left chart is showing the **CPU load** per TaskManager. We've added this metric to the dashboard, since this is the metric used by Kubernetes to decide on the replica count of the TaskManager deployment. We can see that as soon as a certain CPU load is reached, additional TaskManagers are started.
```

Comment:
```suggestion
- The bottom left chart shows the **CPU load** per TaskManager. We've added this metric to the dashboard because this is what Kubernetes will use to decide on the replica count of the TaskManager deployment. You can see that, as soon as a certain CPU load is reached, additional TaskManagers are started.
```